### PR TITLE
#Network level hints 1: Comm split type: minimum communicator size hint

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
This PR contains the minimum communicator size hint for tree topology. When the user supplies a hint indicating the minimum communicator size, subcommunicators are created such that each contains atleast as many processes are the hint. The algorithm works as follows:

1. step 1 indexes leaf nodes of tree topology from left to right and identifies the leaf node (host node) each process is attached to. 
2. The index leaf node for each process is communicated to all processes such that every process now has its own index and the number of processes attached with each leaf node in the tree.
3. Unique color for comm split type is computed by counting how many processes are attached to each leaf node from left to right, in the indexed leaf node array.

Subcommunicator creation first works at network level, splitting further at node level. Once the subcommunicator is identified to be local to a node, assuming that the hint is 3 and the number of processes per node is 4, the algorithm tries to split further within the node to see if it can get closer to 3 per subcommunicator.  If the hint is 5 and number of processes on the node is 4, the subcommunicator is created such that it covers two nodes, instead of one.

### CODE REVIEW

**Comment by KavithaTipturMadhu** [comment]
@raffenet Could you please take a look at this PR?

**Comment by raffenet** [comment]
I realize this wasn't addressed in the previous series, but I'm going to need more than 1-line commit messages to understand what's going on here...

**Comment by KavithaTipturMadhu** [comment]
@raffenet Sorry about that, I have added a description in the comment section here, since the PR is rebased with the netloc PR, you only need to look at the last 3 patches. I can discuss the algorithm in person if the description sounds cryptic.

**Comment by raffenet** [comment]
Thanks, that is what I was looking for. Can you add it to the git history, as well?

File: src/mpi/init/netloc_util.c [hunk 1]
```diff
@@ -402,6 +402,79 @@ int MPIR_Netloc_get_network_end_point(MPIR_Netloc_network_attributes network_att
...
+                                           netloc_node_t * network_endpoint,
+                                           int *index, int *num_leaf_nodes)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int max_level = 0;
+    netloc_node_t **traversal_order;
+    int start_index, end_index, i, j, node_index, num_nodes;
+
+    traversal_order =
+        (netloc_node_t **) MPL_malloc(sizeof(netloc_node_t *) * topology->num_nodes, MPL_MEM_OTHER);
```
Comment (raffenet) [inline src/mpi/init/netloc_util.c]: Let's check that we were able to allocate memory here.

Comment (KavithaTipturMadhu) [reply src/mpi/init/netloc_util.c]: I've updated the PR, there are a couple of other places that use malloc, I'll create a refactoring patch for them.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
@raffenet I have labeled all PRs to be reviewed with numbers, this one goes first.

**Comment by raffenet** [comment]
OK, I am starting with this PR now. I still have the same question about the commit order. This series starts with a commit containing fixups, and at least some of the diff looks like it belongs with the commit that follows (https://github.com/pmodels/mpich/pull/3193/commits/bfdb8bccf429f078eed886b6c0ced6ee1974f514).

As for the minimum size hint, does this have to be used in conjunction with some other hint to be meaningful? Trying to understand the usage.

**Comment by KavithaTipturMadhu** [comment]
The first two commits are more bug fixes than cleanup, the next commit adds a utility function used in comm_split_type to support the minimum subcommunicator size hint.  The hint indicates that the parent communicator is split such that each subcommunicator has at least as many processes as the user indicates. This hint needn't be used with other hints. The split function tries to place processes that share switches of a fat tree closer to each other.

File: src/include/netloc_util.h [hunk 2]
```diff
@@ -40,4 +42,9 @@ int MPIR_Netloc_get_switches_at_level(netloc_topology_t netloc_topology,
                                       MPIR_Netloc_network_attributes attributes, int level,
                                       netloc_node_t *** switches_at_level, int *switch_count);
 
+int MPIR_Netloc_get_hostnode_index_in_tree(MPIR_Netloc_network_attributes attributes,
```
Comment (raffenet) [inline src/include/netloc_util.h]: This must belong in the next commit, right? This function doesn't exist yet.

Comment (KavithaTipturMadhu) [reply src/include/netloc_util.h]: Oh sorry, i seem to have mixed this commit with the next one, will move it to the next commit.

Comment (KavithaTipturMadhu) [reply src/include/netloc_util.h]: Fixed this, along with a couple of typos in commit messages.

**Comment by KavithaTipturMadhu** [comment]
@raffenet Can you take a look at this PR again?

**Comment by raffenet** [comment]
This looks pretty good to me. I'll run the tests again but am prepared to signoff.
test:jenkins/ch3/most
test:jenkins/ch4/most

**Pull request merged**